arch/x86/kvm/pkvm/pkvm.c

```
@@ -620,6 +620,56 @@ static void pkvm_update_exception_bitmap(struct kvm_vcpu *vcpu)
 	kvm_x86_call(update_exception_bitmap)(vcpu);
 }
 
+static int pkvm_set_msr(struct kvm_vcpu *vcpu, u32 index, u64 data)
+{
+	if (pkvm_is_protected_vcpu(vcpu)) {
+		u64 cur_data;
+		int ret;
+
+		if (WARN_ON(kvm_vcpu_has_run(vcpu)))
+			return -EPERM;
+
+		/*
+		 * For simplicity and security, allow the host to change
+		 * initial values of those MSRs (or individual bits in MSRs)
+		 * that are currently tweaked by crosvm, and only those.
+		 * The allowed set can be extended as needed.
+		 */
+		switch (index) {
+		case MTRRphysBase_MSR(0) ... MSR_MTRRfix4K_F8000:
+		case MSR_MTRRdefType:
+			break;
+		case MSR_IA32_MISC_ENABLE:
+			if (data & ~(MSR_IA32_MISC_ENABLE_FAST_STRING |
+				     MSR_IA32_MISC_ENABLE_PEBS_UNAVAIL |
+				     MSR_IA32_MISC_ENABLE_BTS_UNAVAIL))
+				return -EPERM;
+
+			/*
+			 * vPMU is not supported by pKVM yet. Don't trick the pVM
+			 * that it is.
+			 */
+			data |= MSR_IA32_MISC_ENABLE_PEBS_UNAVAIL |
+				MSR_IA32_MISC_ENABLE_BTS_UNAVAIL;
+			break;
+		default:
+			/*
+			 * Allow the host to set an MSR to a value to which it is
+			 * already set by pKVM anyway, i.e. if setting is a no-op.
+			 */
+			ret = kvm_msr_read(vcpu, index, &cur_data);
+			if (ret)
+				return ret;
+			if (data == cur_data)
+				return 0;
```

I wounder if this will not cause us some issues for multiple reasons:
1) I can imagine that accessing some MSRs have side effects and even reading or writing the same value will give a different effect.
2) there are WO or RO MSRs, so e.g. reading WO register will result with unpredicted data and result with EPERM as read data is different than data that we expect to write

Some example e.g. MSR_IA32_FLUSH_CMD (which BTW I did not encounter to be triggered by host in this context, just giving an example of such MSR)

I didn't encounter such issue in basic tests but perhaps in some different SoC we could get into difficult to debug issues because of that? So perhaps it would be better to explicitly allow it for MSR_IA32_XSS  MSR_STAR MSR_CSTAR MSR_L MSR_KERNEL_GS_BASE, MSR_IA32_SYSENTER_CS MSR_IA32_SYSENTER_ESP MSR_IA32_SYSENTER_EIP? which are all store/configuration and doesn't trigger event/hw commands and warn otherwise? WDYT?

---

Indeed, very good point. Thanks for pointing out.

Yeah, looks like we should just allow setting those 5 MSRs set by crosvm (MSR_STAR, MSR_CSTAR, MSR_KERNEL_GS_BASE, MSR_SYSCALL_MASK, MSR_LSTAR). I assume it shouldn't be a security concern, since the guest is supposed to set up those MSRs on its own, before starting using SYSCALL or SWAPGS instructions; in other words, _any_ initial values of those MSRs are untrusted anyway, since they ultimately specify some addresses in the guest memory, while the entire guest memory contents are initially untrusted anyway.

Probably no need to allow MSR_IA32_XSS. KVM sets it to 0 in kvm_vcpu_reset(), but pKVM does the same anyway, and KVM's kvm_vcpu_reset() doesn't check the return value, so it should be ok if it will just silently return -EPERM and do nothing.

---

Just for the record all MSRs that I've mentioned (not only 5 above) I've saw in pkvm_set_msr default handling either on real hw or when running pKVM in QEMU with some VMs:
- MSR_IA32_XSS (0x00000da0) [BTW agree with your reasoning that it could be skipped]
- MSR_STAR (0xc0000081)
- MSR_LSTAR (0xc0000082)
- MSR_CSTAR (0xc0000083)
- MSR_SYSCALL_MASK (0xc0000084)
- MSR_KERNEL_GS_BASE (0xc0000102)
- MSR_IA32_SYSENTER_CS (0x174)
- MSR_IA32_SYSENTER_ESP (0x175)
- MSR_IA32_SYSENTER_EIP (0x176)

---

Yeah you are right, crosvm sets those as well:
```
/// Configure Model specific registers for long (64-bit) mode.
pub fn set_long_mode_msrs(msrs: &mut BTreeMap<u32, u64>) {
    msrs.insert(crate::msr_index::MSR_IA32_SYSENTER_CS, 0x0);
    msrs.insert(crate::msr_index::MSR_IA32_SYSENTER_ESP, 0x0);
    msrs.insert(crate::msr_index::MSR_IA32_SYSENTER_EIP, 0x0);
```